Source/WebCore/rendering/RenderBlock.cpp

```diff
@@ -154,82 +156,82 @@ class PositionedDescendantsMap {
+                    if ((*it).isFixedPositioned()) {
+                        isNewEntry = descendants->insertBefore(it, positionedDescendant).isNewEntry;
                         return;
                     }
                 }
-                isNewEntry = descendants->appendOrMoveToLast(&positionedDescendant).isNewEntry;
+                isNewEntry = descendants->appendOrMoveToLast(positionedDescendant).isNewEntry;
             };
             ensureLayoutDepentBoxPosition();
         }
 
         if (!isNewEntry) {
-            ASSERT(m_containerMap.contains(&positionedDescendant));
+            ASSERT(m_containerMap.contains(positionedDescendant));
             return;
         }
-        m_containerMap.set(&positionedDescendant, &containingBlock);
+        m_containerMap.set(positionedDescendant, &containingBlock);
     }
 
     void removeDescendant(const RenderBox& positionedDescendant)
     {
-        auto* containingBlock = m_containerMap.take(&positionedDescendant);
+        auto containingBlock = m_containerMap.take(positionedDescendant);
         if (!containingBlock)
             return;
 
-        auto descendantsIterator = m_descendantsMap.find(containingBlock);
+        auto descendantsIterator = m_descendantsMap.find(*containingBlock);
         ASSERT(descendantsIterator != m_descendantsMap.end());
         if (descendantsIterator == m_descendantsMap.end())
             return;
 
         auto& descendants = descendantsIterator->value;
-        ASSERT(descendants->contains(const_cast<RenderBox*>(&positionedDescendant)));
+        ASSERT(descendants->contains(const_cast<RenderBox&>(positionedDescendant)));
 
-        descendants->remove(const_cast<RenderBox*>(&positionedDescendant));
-        if (descendants->isEmpty())
+        descendants->remove(const_cast<RenderBox&>(positionedDescendant));
+        if (descendants->isEmptyIgnoringNullReferences())
             m_descendantsMap.remove(descendantsIterator);
     }
     
     void removeContainingBlock(const RenderBlock& containingBlock)
     {
-        auto descendants = m_descendantsMap.take(&containingBlock);
+        auto descendants = m_descendantsMap.take(containingBlock);
         if (!descendants)
             return;
 
-        for (auto* renderer : *descendants)
+        for (auto& renderer : *descendants)
             m_containerMap.remove(renderer);
     }
     
     TrackedRendererListHashSet* positionedRenderers(const RenderBlock& containingBlock) const
     {
-        return m_descendantsMap.get(&containingBlock);
+        return const_cast<TrackedRendererListHashSet*>(m_descendantsMap.get(containingBlock));
     }
 
 private:
-    using DescendantsMap = HashMap<const RenderBlock*, std::unique_ptr<TrackedRendererListHashSet>>;
-    using ContainerMap = HashMap<const RenderBox*, const RenderBlock*>;
+    using DescendantsMap = WeakHashMap<const RenderBlock, std::unique_ptr<TrackedRendererListHashSet>>;
```

Comment:
I believe that `TrackrdRendererListHashSet` is a single possibly null pointer to a hash table, and it is wasteful to store it in turn in a `unique_ptr`. We should consider fixing this mistake at some point.